scylla/src/transport/locator/mod.rs

```diff
@@ -4,5 +4,448 @@ pub mod replication_info;
+    pub fn ring(&self) -> &TokenRing<Arc<Node>> {
+        &self.replication_data.global_ring
+    }
+
+    /// Gives a list of all nodes in the token ring.
+    pub fn unique_nodes_in_global_ring(&self) -> &[Arc<Node>] {
+        self.replication_data.unique_nodes_in_global_ring()
+    }
+
+    /// Gives a list of all known datacenters.
+    pub fn datacenter_names(&self) -> &[String] {
+        self.datacenters.as_slice()
+    }
+
+    /// Gives a list of all nodes in a specified datacenter ring (which is created by filtering the
+    /// original ring to only contain nodes living in the specified datacenter).
+    pub fn unique_nodes_in_datacenter_ring<'a>(
+        &'a self,
+        datacenter_name: &str,
+    ) -> Option<&'a [Arc<Node>]> {
+        self.replication_data
+            .unique_nodes_in_datacenter_ring(datacenter_name)
+    }
+
+    fn get_simple_strategy_replicas(
+        &self,
+        token: Token,
+        replication_factor: usize,
+    ) -> Replicas<'_> {
+        if replication_factor == 0 {
+            return EMPTY_REPLICAS;
+        }
+
+        if let Some(precomputed_replicas) = self
+            .precomputed_replicas
+            .get_precomputed_simple_strategy_replicas(token, replication_factor)
+        {
+            precomputed_replicas.into()
+        } else {
+            Replicas::from_iter(
+                self.replication_data
+                    .simple_strategy_replicas(token, replication_factor),
+            )
+        }
+    }
+
+    fn get_network_strategy_replicas<'a>(
+        &'a self,
+        token: Token,
+        datacenter: &str,
+        datacenter_replication_factor: usize,
+    ) -> Replicas<'a> {
+        if datacenter_replication_factor == 0 {
+            return EMPTY_REPLICAS;
+        }
+
+        if let Some(precomputed_replicas) = self
+            .precomputed_replicas
+            .get_precomputed_network_strategy_replicas(
+                token,
+                datacenter,
+                datacenter_replication_factor,
+            )
+        {
+            Replicas::from(precomputed_replicas)
+        } else {
+            Replicas::from_iter(self.replication_data.nts_replicas_in_datacenter(
+                token,
+                datacenter,
+                datacenter_replication_factor,
+            ))
+        }
+    }
+}
+
+enum ReplicaSetInner<'a> {
+    Plain(Replicas<'a>),
+
+    // Represents a set of SimpleStrategy replicas that is limited to a specified datacenter.
+    FilteredSimple {
+        replicas: Replicas<'a>,
+        datacenter: &'a str,
+    },
+
+    // Represents a set of NetworkTopologyStrategy replicas that is not limited to any specific
+    // datacenter. The set is constructed lazily, by invoking
+    // `locator.get_network_strategy_replicas()`.
+    MergedNTS {
+        datacenter_repfactors: &'a HashMap<String, usize>,
+        locator: &'a ReplicaLocator,
+        token: Token,
+    },
+}
```

Comment:
I think that what the names mean here is what is done with various `Replicas` sequences in order to create the iterator:
- in Plain, there already is one consecutive sequence that is to be iterated,
- in Filtered, such one sequence will be iterated with a predicate filter,
- in Merged, many sequences will be chained (or "merged") in order to create the iterator.

We _could_ change the names into e.g.:
- Plain -> Sequence
- Filtered -> FilteredSequence
- Merged -> ChainedSequences
...but I believe the current names are better. At most, we could replace `Merged` with `Chained`.

Reply:
> I think that the variant names don't describe well what they are supposed to mean, especially the first two ones.
> 
> `Precomputed` `FilteredByDc` `MergedNts` - I don't have a better suggestion here

In addition to what @wprzytula wrote, `Plain` variant does not necessarily represents a precomputed replica sequence. It is because the `Replicas` container may hold:
- a slice to a precomputed sequence
- `Vec` that contains a sequence that was computed on the fly.